File: python/dsl/v1_0/topology.py
class Topology:
    """
    NDH Topology semantic wrapper.
    Represents node/edge counts in runtime graphs.
    """

    def __init__(self, nodes: int = 0, edges: int = 0):
        self.nodes = nodes
        self.edges = edges

    def reversible_add_node(self) -> None:
        self.nodes += 1

    def reversible_add_edge(self) -> None:
        self.edges += 1

File: rust/v1.0/Rust-StructuralRuntime-v1_0/src/topology.rs
#[derive(Debug, Clone)]
pub struct TopologyEnvelope {
    pub nodes: usize,
    pub edges: usize,
}

impl TopologyEnvelope {
    pub fn new() -> Self {
        Self { nodes: 0, edges: 0 }
    }

    pub fn reversible_add_node(&mut self) {
        self.nodes += 1;
    }

    pub fn reversible_add_edge(&mut self) {
        self.edges += 1;
    }
}
